### PR TITLE
components: Add TextInput

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Description
TextInput is a very complicated, low level text input compontent. It handles tons of different use cases including auto-resizing textareas as well as regular text inputs.

## How has this been tested?
Storybook, ensure everything works

## Types of changes
New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by sarayourfriend** [comment]
@diegohaz Could you give this a preliminary review before I turn it into a reviewable PR? I've added some aria labels and stuff and tested using the screen reader but I want to make sure I've caught everything. One thing is that I'm wondering whether we need to add any more a11y helpers for certain interactions like the up/down arrows for number inputs or something.

**Comment by diegohaz** [comment]
It's indeed a very complicated component, especially the numeric features. At first glance, they look like separate components. Why is this a single component instead of something like an `Input` and `NumberInput` that would use `Input` underneath?

I've found some issues. They're all related to the number input:

- The `step` attribute is added even when the input has `type="text"` or is a textarea.
- Pressing <kbd>Shift</kbd> + arrow keys twice doesn't increment the number by 10 (keep <kbd>Shift</kbd> pressed and only release the arrow keys).
- Dragging steppers/arrows feels weird. It keeps incrementing/decrementing for a while even after the cursor is released. When the input has a minimum/maximum value, it feels a little buggy as well.

Besides that, I would like to point out this article: [Why the GOV.UK Design System team changed the input type for numbers](https://technology.blog.gov.uk/2020/02/24/why-the-gov-uk-design-system-team-changed-the-input-type-for-numbers/). I haven't gone very deep into that topic, but that seems something we should take into account. They do mention that `type="number"` should only be used for incrementable numbers (which I think is not a problem for this component), but there are other considerations.

I would try to reduce the scope of this component to a minimum, preferably removing the numeric features, and work on the other features in separate PRs (like one PR for the pure number input, and others for steppers, arrows, shiftStep etc.).

File: packages/components/src/text-input/hooks/use-text-input-state.ts [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+		isTypeNumeric,
+	} );
+
+	const keyboardHandlers = useLatestRef(
+		mergeEventHandlers<
+			KeyboardEvent< HTMLInputElement >,
+			typeof baseKeyboardHandlers,
+			typeof numberKeyboardHandlers
+		>( baseKeyboardHandlers, numberKeyboardHandlers )
+	).current;
```
Comment (diegohaz) [inline packages/components/src/text-input/hooks/use-text-input-state.ts:157]: We can't access a mutable ref's value in the render phase. It should be only inside an effect or event handler.

File: packages/components/src/text-input/hooks/use-text-input.js [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+ * @param {Props} props
+ */
+const useRootEventHandlers = ( {
+	dragHandlersRef,
+	inputRef,
+	isFocused,
+	isTypeNumeric,
+} ) => {
+	const canScroll = isTypeNumeric && isFocused;
+	const dragHandlers = dragHandlersRef?.current;
```
Comment (diegohaz) [inline packages/components/src/text-input/hooks/use-text-input.js:39]: Same.

File: packages/components/src/text-input/text-input-arrows.js [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+ */
+
+/**
+ *
+ * @param {Props}                    props
+ * @param {import('react').Ref<any>} forwardedRef
+ */
+function TextInputArrows( props, forwardedRef ) {
+	const { decrement, dragHandlersRef, increment } = props;
+	const dragHandlers = dragHandlersRef?.current;
```
Comment (diegohaz) [inline packages/components/src/text-input/text-input-arrows.js:42]: Same.

File: packages/components/src/text-input/text-input-steppers.js [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+/**
+ *
+ * @param {Props}                    props
+ * @param {import('react').Ref<any>} forwardedRef
+ * @return {import('react').ReactElement} The text input stepper component.
+ */
+function TextInputSteppers( props, forwardedRef ) {
+	const { decrement, disabled, dragHandlersRef, increment } = props;
+	const dragHandlers = dragHandlersRef?.current;
```
Comment (diegohaz) [inline packages/components/src/text-input/text-input-steppers.js:44]: Same.

**Comment by sarayourfriend** [comment]
Yeah, this PR is way too big, I agree. Unfortunately it will take some time to untangle the numeric handling from the regular text input handling. They're pretty embedded in each other and, to be honest, I don't fully understand everything going on with this component.

**Comment by sarayourfriend** [comment]
Closing in favor of just using the existing `InputControl` component.

**Pull request closed without merging**